### PR TITLE
Document IProvidePartitionKeyFromSagaId

### PR DESCRIPTION
**Pull request by lailabougria**

### CODE REVIEW

File: persistence/azure-table/configuration_code_astp_[3,).partial.md [hunk 1]
```diff
@@ -67,3 +67,10 @@ In case installers are enabled, but there's a need to opt out from creating the
 
 snippet: EnableInstallersConfigurationOptingOutFromTableCreation
 
+### Partitioning
+
+When storing saga's, saga IDs are by default deterministically derived from: the saga data, the correlation property name and the correlation property value. This deterministic ID is used as the partition key.
```
Comment (SeanFeldman) [inline persistence/azure-table/configuration_code_astp_[3,).partial.md:72]:
```suggestion
When storing sagas, saga IDs are by default deterministically derived from the saga data:  the correlation property name and the correlation property value. This deterministic ID is used as the partition key.
```

File: persistence/azure-table/configuration_code_astp_[3,).partial.md [hunk 2]
```diff
@@ -67,3 +67,10 @@ In case installers are enabled, but there's a need to opt out from creating the
...
+To deviate from that, the `IProvidePartitionKeyFromSagaId` interface can be injected into behaviors in the logical pipeline stage to set the partition key.
+
+snippet: BehaviorUsingIProvidePartitionKeyFromSagaId
```
Comment (SeanFeldman) [inline persistence/azure-table/configuration_code_astp_[3,).partial.md:75]: The snippet seems to be missing, CI is failing.

File: persistence/azure-table/configuration_code_astp_[3,).partial.md [hunk 3]
```diff
@@ -67,3 +67,10 @@ In case installers are enabled, but there's a need to opt out from creating the
 
 snippet: EnableInstallersConfigurationOptingOutFromTableCreation
 
+### Partitioning
+
+When storing saga's, saga IDs are by default deterministically derived from: the saga data, the correlation property name and the correlation property value. This deterministic ID is used as the partition key.
+To deviate from that, the `IProvidePartitionKeyFromSagaId` interface can be injected into behaviors in the logical pipeline stage to set the partition key.
```
Comment (SeanFeldman) [inline persistence/azure-table/configuration_code_astp_[3,).partial.md:73]: > To deviate from that

Not sure I follow what do you want to say here. Did you mean 

```suggestion
To set a custom partition key derived from saga ID, the `IProvidePartitionKeyFromSagaId` interface implementation can be injected into behaviors in the logical pipeline stage.
```

**Review by SeanFeldman — CHANGES_REQUESTED** [review]

File: persistence/azure-table/configuration_code_astp_[3,).partial.md [hunk 4]
```diff
@@ -67,3 +67,9 @@ In case installers are enabled, but there's a need to opt out from creating the
 
 snippet: EnableInstallersConfigurationOptingOutFromTableCreation
 
+### Partitioning
```
Comment (danielmarbach) [inline persistence/azure-table/configuration_code_astp_[3,).partial.md:70]: I will try to reword this tomorrow. I think we need to make this clearer what the use case is. the intent was:

- enable partitioning and tx for customers that used previous version of the persistence and have sagas with messages in flight
- simple way to achieve partitioning and tx for conversations when sagas are involved to have the capability to precompute the saga id upfront so that business tx can participate even before the saga persister is at play

Comment (lailabougria) [reply persistence/azure-table/configuration_code_astp_[3,).partial.md:70]: Made some changes, please take a look

File: persistence/azure-table/configuration_code_astp_[3,).partial.md [hunk 5]
```diff
@@ -67,3 +67,15 @@ In case installers are enabled, but there's a need to opt out from creating the
...
+
+When storing sagas, saga IDs are by default deterministically derived from the saga data:  the correlation property name and the correlation property value.
+This deterministic ID is used as the partition key. The partition key is the key to achieving transactionality.
+
+There might be scenario's in which it's desirable to set a custom partition key:
+- When upgrading to the Azure Table Persistence, it's possible make use of the added transactionality-features by setting the partition key for existing sagas that may have in-flight messages
+- To enable business operations to participate in transactions before the saga has been created, the saga ID can be precomputed and set as the partition key
```
Comment (SeanFeldman) [inline persistence/azure-table/configuration_code_astp_[3,).partial.md:77]: It's a single transaction, right?

```suggestion
- To enable business operations to participate in the transaction before the saga has been created, the saga ID can be precomputed and set as the partition key
```

File: persistence/azure-table/configuration_code_astp_[3,).partial.md [hunk 6]
```diff
@@ -67,3 +67,15 @@ In case installers are enabled, but there's a need to opt out from creating the
 
 snippet: EnableInstallersConfigurationOptingOutFromTableCreation
 
+### Partitioning
+
+When storing sagas, saga IDs are by default deterministically derived from the saga data:  the correlation property name and the correlation property value.
+This deterministic ID is used as the partition key. The partition key is the key to achieving transactionality.
+
+There might be scenario's in which it's desirable to set a custom partition key:
+- When upgrading to the Azure Table Persistence, it's possible make use of the added transactionality-features by setting the partition key for existing sagas that may have in-flight messages
```
Comment (SeanFeldman) [inline persistence/azure-table/configuration_code_astp_[3,).partial.md:76]:
```suggestion
- When upgrading to the Azure Table Persistence, it's possible to make use of the added transactional features by setting the partition key for existing sagas that may have in-flight messages
```

**Review by SeanFeldman — COMMENTED** [review]
The sample is [not compiling](https://github.com/Particular/docs.particular.net/pull/5148/checks?check_run_id=1516633233#step:4:442).

**Review by SeanFeldman — APPROVED** [review]

File: Snippets/ASP/ASTP_3/Behaviors.cs [hunk 7]
```diff
@@ -76,4 +78,68 @@ public override async Task Invoke(IIncomingLogicalMessageContext context, Func<T
...
+        {
+            var partitionKeyValue = provideOrderId.OrderId;
+            correlationProperty = new SagaCorrelationProperty("OrderId", partitionKeyValue);
+        }
+
+        await partitionKeyFromSagaId1.SetPartitionKey<OrderSagaData>(context, correlationProperty);
+
+        if (context.Headers.TryGetValue(Headers.SagaId, out var sagaIdHeader))
+        {
+            Log.Info($"Saga Id Header: {sagaIdHeader}");
```
Comment (danielmarbach) [inline Snippets/ASP/ASTP_3/Behaviors.cs:105]: I was wondering if we should remove the log statements and just add comments into the snippet. Do you think they provide value on the configuration page?

File: Snippets/ASP/ASTP_3/Behaviors.cs [hunk 8]
```diff
@@ -76,4 +78,68 @@ public override async Task Invoke(IIncomingLogicalMessageContext context, Func<T
...
+        }
+
+        if (context.Extensions.TryGet<TableInformation>(out var tableInformation))
+        {
+            Log.Info($"Table Information: {tableInformation.TableName}");
```
Comment (danielmarbach) [inline Snippets/ASP/ASTP_3/Behaviors.cs:110]: Same question as above

File: Snippets/ASP/ASTP_3/Behaviors.cs [hunk 9]
```diff
@@ -76,4 +78,68 @@ public override async Task Invoke(IIncomingLogicalMessageContext context, Func<T
...
+        }
+
+        Log.Info($"Found partition key '{context.Extensions.Get<TableEntityPartitionKey>().PartitionKey}' from '{nameof(IProvideOrderId)}'");
```
Comment (danielmarbach) [inline Snippets/ASP/ASTP_3/Behaviors.cs:113]: Same as above

File: persistence/azure-table/configuration_code_astp_[3,).partial.md [hunk 10]
```diff
@@ -67,3 +67,15 @@ In case installers are enabled, but there's a need to opt out from creating the
...
+When storing sagas, saga IDs are by default deterministically derived from the saga data:  the correlation property name and the correlation property value.
+This deterministic ID is used as the partition key. The partition key is the key to achieving transactionality.
+
+There might be scenario's in which it's desirable to set a custom partition key:
+- When upgrading to the Azure Table Persistence, it's possible to make use of the added transactional features by setting the partition key for existing sagas that may have in-flight messages
+- To enable business operations to participate in the transaction before the saga has been created, the saga ID can be precomputed and set as the partition key
+
+To set a custom partition key, the `IProvidePartitionKeyFromSagaId` interface implementation can be injected into behaviors in the logical pipeline stage.
+
+snippet: BehaviorUsingIProvidePartitionKeyFromSagaId
```
Comment (danielmarbach) [inline persistence/azure-table/configuration_code_astp_[3,).partial.md:81]:
```suggestion

### Partitioning and compatibility helpers

In order to atomically store data within the same transaction all data written as part of the same message handling pipeline needs to share the same partition key. When storing sagas, saga IDs are by default deterministically derived from the saga data: the correlation property name and the correlation property value. So for conversations involving a saga as part of the message handling pipeline this determinstically derived ID might be a perfect fit as a partition key. Unfortunately the saga ID assignment happens quite late during the message handling and is not exposed to the user code. This makes it difficult to:

- Support transactional processing for existing sagas that where stored with the previous version of the persister and may have in-flight messages -or-
- Expose the deterministically created saga ID early enough to enable business operations to participate in the transaction before the saga has been created

To support the above scenarios the `IProvidePartitionKeyFromSagaId` interface implementation can be injected into behaviors in the logical pipeline stage.

snippet: BehaviorUsingIProvidePartitionKeyFromSagaId

The component will make sure to

- Derive the table to write the business data and the saga data to by looking for the necessary `TableInformation` or falling back to the default table
- Extract the saga id header if present and populate that header value as the partition key for previously created sagas
- When no header information is available and the compatibility mode is enabled do the necessary secondary index lookups and depending on the configuration even fallback to table scanning  and populate the found saga ID as the partition key. -or-
- When no header information is available and the compatibility mode is disabled precompute the deterministc saga ID based on the provided correlation propert information and populate the precomputed value as the partition key.
```

Comment (danielmarbach) [reply persistence/azure-table/configuration_code_astp_[3,).partial.md:81]: @lailabougria @SeanFeldman what do you think about this one? Might need to grammar tweaks though

File: Snippets/ASP/ASTP_3/Behaviors.cs (hunk 7 above)
Comment (SeanFeldman) [reply Snippets/ASP/ASTP_3/Behaviors.cs:105]: I'd keep it but as a Debug. If anyone is copying the code into production code at least they won't pay the penalty of logging it per message in production and they'll see it in the log when running with VS.

File: Snippets/ASP/ASTP_3/Behaviors.cs (hunk 8 above)
Comment (SeanFeldman) [reply Snippets/ASP/ASTP_3/Behaviors.cs:110]: https://github.com/Particular/docs.particular.net/pull/5148/files#r539924565

File: Snippets/ASP/ASTP_3/Behaviors.cs (hunk 9 above)
Comment (SeanFeldman) [reply Snippets/ASP/ASTP_3/Behaviors.cs:113]: https://github.com/Particular/docs.particular.net/pull/5148/files#r539924565

File: persistence/azure-table/configuration_code_astp_[3,).partial.md (hunk 10 above)
Comment (lailabougria) [reply persistence/azure-table/configuration_code_astp_[3,).partial.md]: I incorporated your suggestion and made some grammar/wording changes @danielmarbach

File: Snippets/ASP/ASTP_3/Behaviors.cs (hunk 7 above)
Comment (lailabougria) [reply Snippets/ASP/ASTP_3/Behaviors.cs]: Modified the logs to debug in the snippets for v3 and v4

**Review by danielmarbach — APPROVED** [review]
